client/pluginmanager/drivermanager/instance.go

```diff
@@ -0,0 +1,443 @@
+}
+
+// handleFingerprint updates the node with the current fingerprint status
+func (i *instanceManager) handleFingerprint(fp *drivers.Fingerprint) {
+	di := &structs.DriverInfo{
+		Attributes:        fp.Attributes,
+		Detected:          fp.Health != drivers.HealthStateUndetected,
+		Healthy:           fp.Health == drivers.HealthStateHealthy,
+		HealthDescription: fp.HealthDescription,
+		UpdateTime:        time.Now(),
+	}
+	i.updateNodeFromDriver(i.id.Name, di)
+
+	// log detected/undetected state changes after the initial fingerprint
+	if i.hasFingerprinted {
+		if i.lastHealthState != fp.Health {
+			i.logger.Info("driver health state has changed", "previous", i.lastHealthState, "current", fp.Health, "description", fp.HealthDescription)
+		}
+	}
+	i.lastHealthState = fp.Health
+
+	// if this is the first fingerprint, mark that we have received it
+	if !i.hasFingerprinted {
+		if i.logger.IsTrace() {
+			i.logger.Trace("initial driver fingerprint", "fingerprint", fp)
+		}
+		close(i.firstFingerprintCh)
+		i.hasFingerprinted = true
+	}
+}
+
+// dispenseTaskEventsCh dispenses a driver plugin and makes an TaskEvents RPC.
+// The TaskEvent chan and cancel func for the RPC is return. The cancel func must
+// be called by the caller to properly cleanup the context
+func (i *instanceManager) dispenseTaskEventsCh() (<-chan *drivers.TaskEvent, context.CancelFunc, error) {
+	driver, err := i.dispense()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	ctx, cancel := context.WithCancel(i.ctx)
+	eventsCh, err := driver.TaskEvents(ctx)
+	if err != nil {
+		cancel()
+		return nil, nil, err
+	}
+
+	return eventsCh, cancel, nil
+}
+
+// handleEvents is the main loop that receives task events from the driver
+func (i *instanceManager) handleEvents() {
+	eventsCh, cancel, err := i.dispenseTaskEventsCh()
+	if err != nil {
+		i.logger.Error("failed to dispense driver", "error", err)
+	}
+
+	var backoff time.Duration
+	var retry int
+	for {
+		time.Sleep(backoff)
+		select {
+		case <-i.ctx.Done():
+			cancel()
+			return
+		case ev, ok := <-eventsCh:
+			if ok {
+				i.handleEvent(ev)
+				continue
+			}
+
+			// if the channel is closed attempt to open a new one
+			newEventsChan, newCancel, err := i.dispenseTaskEventsCh()
+			if err != nil {
+				i.logger.Warn("failed to receive task events, retrying", "error", err, "retry", retry)
+
+				// Calculate the new backoff
+				backoff = (1 << (2 * uint64(retry))) * driverFPBackoffBaseline
+				if backoff > driverFPBackoffLimit {
+					backoff = driverFPBackoffLimit
+				}
+				retry++
+				continue
+			}
+			cancel()
+			eventsCh = newEventsChan
+			cancel = newCancel
+
+			// Reset backoff
+			backoff = 0
+			retry = 0
+
+		}
+	}
+}
+
+// handleEvent looks up the event handler(s) for the event and runs them
+func (i *instanceManager) handleEvent(ev *drivers.TaskEvent) {
+	i.handlersLock.RLock()
+	defer i.handlersLock.RUnlock()
+	for _, handler := range i.handlers[ev.TaskID] {
+		handler(ev)
+	}
```

Comment:
We should log events that do not have a handler. I think WARN unless you think it should be something else.